### PR TITLE
implement totp in py/js/php :0

### PR DESCRIPTION
**Pull request by frosty00**
doing php now

how do we add pyotp to requirements.txt i can't find a requirements.txt?

### CODE REVIEW

**Comment by kroitor** [comment]
> how do we add pyotp to requirements.txt i can't find a requirements.txt?

there isn't ) all deps are here: https://github.com/ccxt/ccxt/blob/master/python/setup.py#L73

File: python/ccxt/base/exchange.py [hunk 1]
```diff
@@ -1668,3 +1669,7 @@ def signMessage(self, message, privateKey):
         message_hash = self.hashMessage(message)
         signature = self.signHash(message_hash[-64:], privateKey[-64:])
         return signature
+
+    def oath(self):
```
Comment (kroitor) [inline python/ccxt/base/exchange.py]: Let's make the twofa keypair its argument? Easier to maintain and override.

**Comment by kroitor** [comment]
> how do we add pyotp to requirements.txt

Instead of adding the deps, i suggest to make them optional, as we did with web3 here:

https://github.com/ccxt/ccxt/blob/master/python/ccxt/base/exchange.py#L75

Say, if the user has installed `pyotp` and has set `this.twofa` – then it will generate the otp, otherwise it will throw an exception or something. How does that sound?

**Comment by kroitor** [comment]
Also check this: https://github.com/ccxt/ccxt/blob/master/python/ccxt/base/exchange.py#L1468

**Comment by frosty00** [comment]
I disagree with passing the key as an argument. The otp secret should be a property of an exchange object and you should be able to get it with `exchange.oath()`. Another reason this is true is because exchange objects also may have a timedifference attribute and that should be taken into account aswell.

**Comment by kroitor** [comment]
> I disagree with passing the key as an argument. The otp secret should be a property of an exchange object and you should be able to get it with exchange.oath().

Well, if you think of it a bit deeply, then you'll immediately see that there's nothing to disagree on )

We really mean the same thing:

```Python
# pseudocode
def oath(self):
    return self.totp()

def totp(self):
    return Exchange.make_totp_code_from_totp_key(self.twofa)

@staticmethod
def make_totp_code_from_totp_key(key):
    # you can call this method whatever you like    
    if key is None:
        throw ...
    try:
        result = the_otp_package.the_method_to_generate_it(key)
    except not_installed as e: # pseudocode
        throw ...
    return result
```

Or do you still disagree?

**Comment by frosty00** [comment]
I think we can rewrite this in all three languages so the only requirement is the `hmac` function that already exists.

**Comment by kroitor** [comment]
@frosty00 that would be awesome... because traders are often suspicious with the added dependencies.

**Pull request merged**